src/sync/index.js

```diff
@@ -403,111 +408,126 @@ async function sync(db){
-    if(!oracle){
+async function updateOracleBalance(oracleAddress, topicSet, db, resolve){
+  var oracle;
+  try {
+    oracle = await db.Oracles.findOne({address: oracleAddress});
+    if (!oracle){
+      console.error(`Error: find 0 oracle ${oracleAddress} in db to update`);
       resolve();
       return;
     }
-    // related topic should be updated
-    topicSet.add(oracle.topicAddress);
-    if(oracle.token === 'QTUM'){
-      // centrailized
-      const contract = qclient.Contract(oracleAddress, Contracts.CentralizedOracle.abi);
-      contract.call('getTotalBets',{ methodArgs: [], senderAddress: senderAddress})
-        .then(
-          (value)=>{
-            let balances = _.map(value[0].slice(0, oracle.options.length), (balance_BN) =>{
-              return balance_BN.toJSON();
-            });
-            db.Oracles.updateOne({address: oracleAddress}, { $set: { amounts: balances } }, function(err, res){
-              if(err){
-                console.error(`Error: ${err.message}`);
-              }
-              resolve();
-            });
-          }
-        );
-    }else{
-      // decentralized
-      const contract = qclient.Contract(oracleAddress, Contracts.DecentralizedOracle.abi);
-      contract.call('getTotalVotes', { methodArgs: [], senderAddress: senderAddress})
-      .then(
-        (value)=>{
-          let balances = _.map(value[0].slice(0, oracle.options.length), (balance_BN) =>{
-            return balance_BN.toJSON();
-          });
-          db.Oracles.updateOne({address: oracleAddress}, { $set: { amounts: balances } }, function(err, res){
-            if(err){
-              console.error(`Error: ${err.message}`);
-            }
-            resolve();
-          });
-        }
-      );
+  } catch(err){
+    console.error(`Error: update oracle ${oracleAddress} in db, ${err.message}`);
+    resolve();
+    return;
+  }
+
+  // related topic should be updated
+  topicSet.add(oracle.topicAddress);
+  var value;
+  if(oracle.token === 'QTUM'){
+    // centrailized
+    const contract = new Qweb3Contract(QTUM_RPC_ADDRESS, oracleAddress, Contracts.CentralizedOracle.abi);
+    try {
+      value = await contract.call('getTotalBets',{ methodArgs: [], senderAddress: senderAddress});
+    } catch(err){
+      console.error(`Error: getTotalBets for oracle ${oracleAddress}, ${err.message}`);
+      resolve();
+      return;
     }
+  }else{
+    // decentralized
+    const contract = new Qweb3Contract(QTUM_RPC_ADDRESS, oracleAddress, Contracts.DecentralizedOracle.abi);
+    try {
+      value = await contract.call('getTotalVotes', { methodArgs: [], senderAddress: senderAddress});
+    } catch(err){
+      console.error(`Error: getTotalVotes for oracle ${oracleAddress}, ${err.message}`);
+      resolve();
+      return;
+    }
+  }
+
+  let balances = _.map(value[0].slice(0, oracle.options.length), (balance_BN) => {
+    return balance_BN.toJSON();
```

Comment:
value and value[0] needs undefined check. Access undefined[0] or undefined.slice() will throw an error.

or even better _.isEmpty().

Reply:
i think it is fine 4 now unless qweb3 (ethabi) turns unexpected result without error